File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& vec) {
        // set<vector<int>> st;
        // for(int i=0;i<nums.size();i++){
        //     set<int> hashset;
        //     for(int j=i+1;j<nums.size();j++){
        //         int third=-(nums[i]+nums[j]);
        //         if(hashset.find(third)!=hashset.end()){
        //             vector<int> temp={nums[i],nums[j],third};
        //             sort(temp.begin(),temp.end());
        //             st.insert(temp);
        //         }
        //         hashset.insert(nums[j]);
        //     }
        // }
        // vector<vector<int>> res(st.begin(),st.end());
        // return res;

        // vector<vector<int>> vec;
        // sort(nums.begin(),nums.end());
        // for(int i=0;i<nums.size();i++){
        //     if(i>0 && nums[i]==nums[i-1]) continue;
        //     int j=i+1,k=nums.size()-1;
        //     while(j<k){
        //         if((nums[i]+nums[j]+nums[k])==0){
        //             vector<int> temp={nums[i],nums[j],nums[k]};
        //             vec.push_back(temp);
        //             j++,k--;
        //             while(j<k && nums[j]==nums[j-1]) j++;
        //             while(j<k && nums[k]==nums[k+1]) k--;
        //         }
        //         else if((nums[i]+nums[j]+nums[k])<0){
        //             j++;
        //         }
        //         else{
        //             k--;
        //         }
        //     }
        // }
        // return vec;

        set<vector<int>> res;
        sort(vec.begin(),vec.end());
        for(int i=0;i<vec.size()-2;i++){
            int v=-vec[i];
            int l=i+1,r=vec.size()-1;
            while(l<r){
                int sum=vec[l]+vec[r];
                if(sum==v){
                    res.insert({vec[i],vec[l],vec[r]});
                    l++,r--;
                }
                else if(sum<v){
                    l++;
                }
                else r--;
            }
        }
        vector<vector<int>> out;
        out.insert(out.end(),res.begin(),res.end());
        return out;
    }
};

File: 0015-3sum/0015-3sum.py
# class Solution:
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         a=sorted(nums)
#         res=[]
#         for i in range(len(nums)-2):
#             for j in range(i+1,len(nums)-1):
#                 for k in range(j+1,len(nums)):
#                     b=[]
#                     if(nums[i]+nums[j]+nums[k]==0):
#                         b.append(nums[i])
#                         b.append(nums[j])
#                         b.append(nums[k])
#                         res.append(b)
#         ary=[]
#         for i in res:
#             ary.append(sorted(i))

#         unique_elements = set()
#         result = []
    
#         for inner_list in ary:
#             inner_tuple = tuple(inner_list)
#             if inner_tuple not in unique_elements:
#                 unique_elements.add(inner_tuple)
#                 result.append(inner_list)
#         return result

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        
        for i in range(n - 2):
            # Skip duplicates for i
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            left = i + 1
            right = n - 1
            
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if total == 0:
                    res.append([nums[i], nums[left], nums[right]])
                    # Skip duplicates for left and right pointers
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1
        
        return res

         

